docs/reference/extensions/env-injector-extension.rst

```diff
@@ -0,0 +1,126 @@
+
+- *Global* environment variables are visible to all apps in the snap that use
+  env-injector.
+- *Local* environment variables are visible only to apps with env-ibjector that the user
+  specifically targets.
+
+At a high level, the extension has the following processes and rules during build and
+runtime.
+
+
+Environment variable order of precedence
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Environment variables are processed in the following order:
+
+1. Variables extracted from the global ``.env`` file.
+2. Variables extracted from the local ``.env`` file.
+3. Variables from global snap config.
+4. Variables from local snap config.
+
+
+App build process
+~~~~~~~~~~~~~~~~~
+
+.. TODO: Put a link to exporter program repository
+
+For each app that uses the extension:
+
+1. The app responsible for processing environment variables (the exporter program) is
+   added to its `command chain
+   <https://snapcraft.io/docs/snapcraft-yaml-schema#p-21225-command-chain-47>`_.
+2. The app's name is taken either from its internal name or an :ref:`environment alias
+   <use-the-env-injector-give-app-alias>`.
+
+
+Runtime process
+~~~~~~~~~~~~~~~
+
+When the snap runs, for each app that uses the extension, the exporter program:
+
+1. Runs before the app is executed.
+2. Makes a request to the `snapd API <https://snapcraft.io/docs/using-the-api>`_
+   through the snapd Unix socket.
+3. Reads the available environment variables or paths to the ``.env`` files.
+4. Translates the snap options into environment variables.
+5. If available, loads the ``.env`` files and sets the listed variables.
+6. Sets all converted environment variables.
+7. Executes the app's command.
+
+
+.. _env-injector-naming-rules:
+
+Snap option naming and rules
+----------------------------
+
+Because there's a chance of conflict and collision between the names of snap options and
+environment variables, this extension is governed by naming rules.
+
+Snap options map to environment variables:
+
+.. list-table::
+    :header-rows: 1
+
+    * - Type
+      - Snap configuration option
+      - Environment variable
+    * - Global
+      - ``env.<key>=<value>``
+      - ``<key>=<value>``
+    * - Local
+      - ``apps.<app>.env.<key>=<value>``
+      - ``<key>=<value>``
+
+When mapped, key names are sanitized before being converted into environment variable
+names:
+
+.. list-table::
+    :header-rows: 1
+
+    * - Character in snap option
+      - Result in environment variable
+    * - Lowercase letters
+      - Uppercase letters
+    * - Number not at key start
+      - Number
+    * - Hyphen not at key start or end
+      - Underscore
+
+.. tip::
+
+    The first column is the rule set for snap option names. Any character not detailed
+    in it is invalid in a snap option name.
+
+
+Environment file paths
+----------------------
+
+Environment (``.env``) files can pass variables to the snap with the special ``envfile``
+option:
```

Comment:
Worth noting here that the ``envfile`` must be readable from within the snap?

Reply:
I wasn't happy with this section, so I rewrote it: f296e3f

This comes with the proviso that I'm probably way off with the phrase "user generated". I presume such a thing is impossible, but it leads me to ask what the purpose of reading env files inside the snap are, if the user is limited to what the snap authors originally included. Alternative runtime modes?

Reply:
I guess theoretically one could pack different env files in the snap for that and point `envfile` to `/snap/<my-snap>/current/some_file.env`, but I don't think that's the intended use case and I'd shy away from documenting that right now.. If you want to be more general you could probably get away with "user-configured", but tbh it's much of a muchness at this point and I'm happy with it as-is.

Reply:
Tweaked the heading: [2ea902a](https://github.com/canonical/snapcraft/pull/5212/commits/2ea902a7e86a04dab4b85ec86aa3d13c9d111e22)

Overlapping with the other thread – if we're unsure about implementation, let's leave out user-generated envs for now.